server/server_test.go

```diff
@@ -510,3 +514,220 @@ func TestTranslateGrpcCookieHeader(t *testing.T) {
 	})
 
 }
+
+// Ensure CSRF key is available when set in Secret
+func Test_CSRFKeyIsAvailableFromSettings(t *testing.T) {
+	s := fakeServer(true, false)
+	assert.False(t, s.DisableCsrf)
+	assert.NotNil(t, s.settings.CsrfKey)
+	assert.Len(t, s.settings.CsrfKey, 32)
+}
+
+func Test_CSRFProtection_Disabled(t *testing.T) {
+	s := fakeServer(true, true)
+	cancelInformer := test.StartInformer(s.projInformer)
+	defer cancelInformer()
+	port, err := test.GetFreePort()
+	assert.NoError(t, err)
+	metricsPort, err := test.GetFreePort()
+	assert.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go s.Run(ctx, port, metricsPort)
+	defer func() { time.Sleep(3 * time.Second) }()
+
+	err = test.WaitForPortListen(fmt.Sprintf("127.0.0.1:%d", port), 10*time.Second)
+	assert.NoError(t, err)
+
+	// ALlow server a few milliseconds to settle
+	time.Sleep(1 * time.Second)
+
+	client := http.Client{}
+	serverURL := fmt.Sprintf("http://127.0.0.1:%d", port)
+
+	// With CRSF protection disabled, server should neither send CSRF header nor cookie
+	{
+		cookieSent := false
+		req, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/session/userinfo", serverURL), nil)
+		assert.NoError(t, err)
+		resp, err := client.Do(req)
+		assert.NoError(t, err)
+		assert.Equal(t, 200, resp.StatusCode)
+		csrftoken := resp.Header.Get("X-CSRF-Token")
+		assert.Empty(t, csrftoken)
+		for _, cookie := range resp.Cookies() {
+			if cookie.Name == "_gorilla_csrf" {
+				cookieSent = true
+			}
+		}
+		assert.False(t, cookieSent)
+	}
+
+	// With CSRF protection disabled, these requests should get through to the API
+	// All HTTP status codes except 403 are fine for us
+	for _, method := range []string{"POST", "DELETE", "PUT"} {
+		req, err := http.NewRequest(method, fmt.Sprintf("%s/api/v1/repositories", serverURL), bytes.NewBuffer([]byte("{}")))
+		assert.NoError(t, err)
+		resp, err := client.Do(req)
+		assert.NoError(t, err)
+		assert.NotNil(t, resp)
+		assert.NotEqual(t, 403, resp.StatusCode)
+	}
+}
+
+func TestCsrfProtection(t *testing.T) {
+	s := fakeServer(false, false)
+	cancelInformer := test.StartInformer(s.projInformer)
+	defer cancelInformer()
+	port, err := test.GetFreePort()
+	assert.NoError(t, err)
+	metricsPort, err := test.GetFreePort()
+	assert.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go s.Run(ctx, port, metricsPort)
+	defer func() { time.Sleep(3 * time.Second) }()
+
+	err = test.WaitForPortListen(fmt.Sprintf("127.0.0.1:%d", port), 10*time.Second)
+	assert.NoError(t, err)
+
+	// ALlow server a few milliseconds to settle
+	time.Sleep(1 * time.Second)
```

Comment:
Sometimes, the fake server started in these tests need some time to settle or it will return HTTP error 503 on every request. It failed sometimes locally with this behaviour, I hope it doesn't happen in the future.

Reply:
I feel like we should have non-time based set-up and clean-up. Don't make any changes until we've discussed the bigger picture on this please.

Reply:
Alright, just catch me on Slack or let's have a Zoom call, whatever you prefer